### ISSUE TITLE
read error detection jobs failed

### ISSUE THREAD

**Issue by narechan** [marbl/canu#269]
Hi,

I've been struggling with the the assembly portion of a canu run that is giving me the following error:

```
> ....
> RED job 14411 from read   2121325 to read   2121480 -   1.512 GB for     156 reads -   0.013 GB for   1056691 olaps -   0.005 GB for evidence
> RED job 14412 from read   2121481 to read   2121640 -   1.503 GB for     160 reads -   0.013 GB for   1050648 olaps -   0.005 GB for evidence
> RED job 14413 from read   2121641 to read   2121785 -   1.504 GB for     145 reads -   0.013 GB for   1052042 olaps -   0.004 GB for evidence
> RED job 14414 from read   2121786 to read   2121944 -   1.507 GB for     159 reads -   0.013 GB for   1052990 olaps -   0.005 GB for evidence
> RED job 14415 from read   2121945 to read   2121996 -   0.513 GB for      52 reads -   0.004 GB for    358327 olaps -   0.002 GB for evidence
> --
> -- 1 read error detection jobs failed:
> --   job /isilon/seq/apurva/analysis/canu/man/mat_10000_09292016_3/unitigging/3-overlapErrorAdjustment/0000.red FAILED.
> --
> ================================================================================
> Don't panic, but a mostly harmless error occurred and canu failed.
> 
> canu failed with 'failed to detect errors in reads.  Made 2 attempts, jobs still failed'.
> 
```

here is the command i'm issuing:

```
canu -assemble -d mat_10000_09292016_3 -p mat_10000_09292016_3 genomeSize=20000000 gridOptionsExecutive="-l mem_free=32g" gridEngineMemoryOption="-l m_mem_free=MEMORY" corOutCoverage=10000 java=/usr/bin/java -pacbio-corrected mat_10000_09292016.trimmedReads.fastq
```

This particular sample has a lot of contamination and can almost be considered metagenomic. We are trying to get to some of the low coverage stuff, hence the high value for corOutCoverage. This has lead to a pretty large corrected/trimmed read file:

```
-rw-r--r-- 1 apurva seqbh  29G Oct  5 20:32 mat_10000_09292016.trimmedReads.fastq
```

One note that may be relevant is that I used an older version of canu (1.1) to do the read correction and trimming. I've since switched to the most recent unreleased version of canu because of feedback given on this issue:

https://github.com/marbl/canu/issues/199

But I'm still getting the same error, unfortunately.

Here is my current version:

$ canu --version
Canu v1.3 (+218 commits) r7834 0049b3bfc0dbe2f1f6d2dd4d96763ad9a4826202.

Please let me know if you need any further information.

Thanks for your help,
Apurva Narechania.

**Comment by brianwalenz** [comment]
Interesting.  It took me a while to figure out what happened.  With 14415 jobs, our four-digit job ID overflowed.  Job 10000 was assigned the ID of '0000', which doesn't exist.  Also, job 10001 was assigned an id of "0001", which already existed, and so job 10001 never ran.

I don't have a fix io detect and correct the problem.  You need to increase redMemory (currently 2gb) to something larger.  I suggest setting both:
```
redMemory=8
oeaMemory=4
```
Before restarting, delete the entire 3-overlapErrorAdjustment directory.  There's nothing to salvage in there, and the old partitioning will just confuse things.

**Comment by narechan** [comment]
Ha! I see. Thanks.

The assembly now proceeded through that stage presumably with fewer jobs. But I am getting a unitigger error now. I'm guessing it may have to do with how many tigs I'm carrying into the process? It's failing at 'PLACE CONTAINED READS'. I've copied the complete error message below. This is the command I am now running:

canu -assemble -d mat_10000_09292016_3 -p mat_10000_09292016_3 redMemory=8 oeaMemory=4 genomeSize=20000000 gridOptionsExecutive="-l mem_free=32g" gridEngineMemoryOption="-l m_mem_free=MEMORY" corOutCoverage=10000 java=/usr/bin/java -pacbio-corrected mat_10000_09292016.trimmedReads.fastq

Error message:


Graph  error threshold  = 0.075 (7.500%)
Max    error threshold  = 0.075 (7.500%)

Minimum overlap length = 500 bases

number of threads     = 4 (command line)


==> LOADING AND FILTERING OVERLAPS.

ReadInfo()-- Using 2121996 reads, no minimum read length used.

OverlapCache()-- limited to 15360MB memory (user supplied).
PHYS_PAGES = 33032620
PAGE_SIZE  = 4096
MEMORY     = 135301611520
OverlapCache()--      24MB for read data.
OverlapCache()--      32MB for best edges.
OverlapCache()--      56MB for unitig layouts.
OverlapCache()--       0MB for tigs.
OverlapCache()--      16MB for id maps.
OverlapCache()--     129MB for error profiles.
OverlapCache()--      24MB for overlap cache pointers.
OverlapCache()--      56MB for overlap cache initial bucket.
OverlapCache()--      64MB for overlap cache thread data.
OverlapCache()--       8MB for number of overlaps per read.
OverlapCache()--       0MB for other processes.
OverlapCache()-- ---------
OverlapCache()--     282MB for data structures (sum of above).
OverlapCache()--   15077MB available for overlaps.

OverlapCache()-- Loading number of overlaps per read.
OverlapCache()--  Initial guess at _maxPer=465 (max of 15140) from (memLimit=15810008021 - memUsed=0) / (numReads=2121996 * sizeof(OVL)=16)
OverlapCache()-- _maxPer=    465 (numBelow=590943 numEqual=170 numAbove=1530883 totalLoad=823743134 -- 0 + 823743134 = 13179890144 <? 15810008021
OverlapCache()--                 (2508 MB free, adjust by 107)
OverlapCache()-- _maxPer=    572 (numBelow=604375 numEqual=77 numAbove=1517544 totalLoad=986734846 -- 0 + 986734846 = 15787757536 <? 15810008021
OverlapCache()--                 (21 MB free, adjust by 0)

OverlapCache()-- blockSize        = 15439460 (235MB)

OverlapCache()-- _maxPer          = 572 overlaps/reads
OverlapCache()-- numBelow         = 604375 reads (all overlaps loaded)
OverlapCache()-- numEqual         = 77 reads (all overlaps loaded)
OverlapCache()-- numAbove         = 1517544 reads (some overlaps loaded)
OverlapCache()-- totalLoad        = 986734846 overlaps ( 15.95%)

OverlapCache()-- availForOverlaps = 15077MB
OverlapCache()-- totalMemory      = 0MB for organization
OverlapCache()-- totalMemory      = 15056MB for overlaps
OverlapCache()-- totalMemory      = 15056MB used

OverlapCache()-- Loading: overlaps processed         5707 (000.00%) loaded          572 (000.00%) droppeddupe            0 (000.00%) (at read iid 1)
OverlapCache()-- Loading: overlaps processed   2924331948 (047.27%) loaded    461990825 (007.47%) droppeddupe       230122 (000.00%) (at read iid 1001447)
OverlapCache()-- Loading: overlaps processed   5840764883 (094.42%) loaded    930810686 (015.05%) droppeddupe       392155 (000.01%) (at read iid 2002676)
OverlapCache()-- Loading: overlaps processed   6186218544 (100.00%) loaded    986594596 (015.95%) droppeddupe       410156 (000.01%)

BestOverlapGraph()-- allocating best edges (64MB)
BestOverlapGraph()-- analyzing 2121996 reads for best contains, with 4 threads.
BestOverlapGraph()-- analyzing 2121996 reads for best edges, with 4 threads.
BestOverlapGraph()-- removing suspicious reads from graph, with 4 threads.
BestOverlapGraph()-- marked 2825 reads as suspicious.
BestOverlapGraph()-- analyzing 2121996 reads for best contains, with 4 threads.
BestOverlapGraph()-- analyzing 2121996 reads for best edges, with 4 threads.
BestOverlapGraph()-- analyzing best edges to find useful edge error rate
BestOverlapGraph()-- analyzing 2121996 reads for best contains, with 4 threads.
BestOverlapGraph()-- analyzing 2121996 reads for best edges, with 4 threads.
BestOverlapGraph()-- removing suspicious edges from graph, with 4 threads.
BestOverlapGraph()-- analyzing 2121996 reads for best contains, with 4 threads.
BestOverlapGraph()-- analyzing 2121996 reads for best edges, with 4 threads.
BestOverlapGraph()-- detecting spur reads.
BestOverlapGraph()-- detected 13188 spur reads.
BestOverlapGraph()-- analyzing 2121996 reads for best contains, with 4 threads.
BestOverlapGraph()-- analyzing 2121996 reads for best edges, with 4 threads.
BestOverlapGraph()-- removing best edges for contained reads.

==> BUILDING GREEDY TIGS.

breakSingletonTigs()-- Removed 57509 singleton tigs; reads are now unplaced.

==> PLACE CONTAINED READS.

computeErrorProfiles()-- Computing error profiles for 59126 tigs, with 4 threads.

Failed with 'Segmentation fault'

Backtrace (mangled):

/sonas-hs/mccombie/hpc/home/apurva/bin/canu/Linux-amd64/bin/bogart(_Z17AS_UTL_catchCrashiP7siginfoPv+0x27)[0x4ecec7]
/lib64/libpthread.so.0[0x393080f710]
/sonas-hs/mccombie/hpc/home/apurva/bin/canu/Linux-amd64/bin/bogart[0x4d83b3]
/sonas-hs/mccombie/hpc/home/apurva/bin/canu/Linux-amd64/bin/bogart[0x4d6ae5]
/sonas-hs/mccombie/hpc/home/apurva/bin/canu/Linux-amd64/bin/bogart[0x404534]
/lib64/libc.so.6(__libc_start_main+0xfd)[0x393001ed5d]
/sonas-hs/mccombie/hpc/home/apurva/bin/canu/Linux-amd64/bin/bogart[0x4033b9]

Backtrace (demangled):

[0] /sonas-hs/mccombie/hpc/home/apurva/bin/canu/Linux-amd64/bin/bogart::AS_UTL_catchCrash(int, siginfo*, void*) + 0x27  [0x4ecec7]
[1] /lib64/libpthread.so.0() [0x393080f710]
[2] /sonas-hs/mccombie/hpc/home/apurva/bin/canu/Linux-amd64/bin/bogart() [0x4d83b3]
[3] /sonas-hs/mccombie/hpc/home/apurva/bin/canu/Linux-amd64/bin/bogart() [0x4d6ae5]
[4] /sonas-hs/mccombie/hpc/home/apurva/bin/canu/Linux-amd64/bin/bogart() [0x404534]
[5] /lib64/libc.so.6::(null) + 0xfd  [0x393001ed5d]
[6] /sonas-hs/mccombie/hpc/home/apurva/bin/canu/Linux-amd64/bin/bogart() [0x4033b9]

GDB:

**Comment by brianwalenz** [comment]
I think I've discovered what is causing this to fail, but haven't worked out a solution yet.  A potential work around is to increase the memory allowed for the bogart command.  You seem to have 32gb, so try batMemory=28.  The more the better.

**Comment by narechan** [comment]
Ok, so I managed to get this to assemble in two ways: 1. removing high concentration / obvious bacterial contaminants with blasr (which removed nearly half of my sequence) and running the command as above with no batMemory specified, and 2. using all the data with bogart running batMemory=500. I ran batMemory with 28 as you suggested, and 100 after that. When it failed at 100, I jacked it up to 500G. Since this probably isn't practical for many people, a software solution would probably be preferable!

Thanks for your attention to this!
